lib/puppet/util/execution.rb

```diff
@@ -319,6 +322,15 @@ class << self
   # @api private
   #
   def self.execute_posix(command, options, stdin, stdout, stderr)
+    # NOTE: The reason we do not put this check in 'execute' is because
+    # on Windows, cwd could be a UNC path. In this case, there's a
+    # chance that File.directory?(cwd) could return false when the UNC
+    # path exists, which is incorrect. See PUP-6919.
+    cwd = options[:cwd] || Dir.pwd
+    unless File.directory?(cwd)
```

Comment:
Might be better to catch the ENOENT error thrown by Kernel.exec? This way, we ensure that all knowledge of cwd handling is in the Kernel.exec call since that's where we cd to our cwd. Not sure about this one.